lib/sdoc/search_index.rb

```diff
@@ -0,0 +1,140 @@
+require "base64"
+require "nokogiri"
+
+module SDoc::SearchIndex
+  extend self
+
+  def generate(rdoc_modules)
+    # RDoc duplicates RDoc::MethodAttr instances when modules are aliased by
+    # assigning to a constant. For example, `MyBar = Foo::Bar` will duplicate
+    # all of Foo::Bar's RDoc::MethodAttr instances.
+    rdoc_objects = rdoc_modules + rdoc_modules.flat_map(&:method_list).uniq
+
+    bigram_sets = rdoc_objects.map { |rdoc_object| derive_bigrams(rdoc_object.full_name) }
+    bigram_bit_positions = compile_bigrams(bigram_sets)
+    bit_weights = compute_bit_weights(bigram_bit_positions)
+    entries = []
+
+    rdoc_objects.zip(bigram_sets) do |rdoc_object, bigrams|
+      entries << [
+        generate_fingerprint(bigrams, bigram_bit_positions), # Fingerprint
+        0.1 / Math.sqrt(rdoc_object.full_name.length), # Tie-breaker bonus per matching bigram
+        rdoc_object.path, # URL
+      ]
+
+      if rdoc_object.is_a?(RDoc::ClassModule)
+        entries.last << name_for(rdoc_object) # Class name
+        entries.last << nil # Method name
+      else
+        entries.last << name_for(rdoc_object.parent) # Class name
+        entries.last << name_for(rdoc_object) # Method name
+
+        # Reduce tie-breaker bonus in proportion to method name length. This
+        # prioritizes modules before methods, and short methods of long modules
+        # before long methods of short modules. For example, when searching for
+        # "find_by", this prioritizes ActiveRecord::FinderMethods#find_by before
+        # ActiveRecord::Querying#find_by_sql.
+        entries.last[1] *= 0.95 ** rdoc_object.name.length
```

Comment:
Can't we favor the position of objects with a more "sophisticated" description as well ?

For instance, I always need to check `ActionController::Rendering#render`'s documentation but there's a dozen `#render` methods that I don't care about, some of them are better ranked but have a very succinct description.

Reply:
That's an interesting idea!  It's tricky though because the bonus is per bigram instead of per entry.  And in cases like [`ActiveRecord::FinderMethods#find_by`][] vs [`ActiveRecord::Querying#find_by_sql`][], we definitely want "find_by" to find `#find_by`, but `#find_by_sql` has much longer documentation.

I've pushed a commit that tries to balance the constraints, but it might be overtuned to the specific examples I used.  Please play around with it and let me know if you encounter any odd results!

[`ActiveRecord::FinderMethods#find_by`]: https://api.rubyonrails.org/v7.1/classes/ActiveRecord/FinderMethods.html#method-i-find_by
[`ActiveRecord::Querying#find_by_sql`]: https://api.rubyonrails.org/v7.1/classes/ActiveRecord/Querying.html#method-i-find_by_sql